src/dragen_align_pa/stages.py

```diff
@@ -0,0 +1,345 @@
+            api_root=ICA_REST_ENDPOINT,
+            sg_name=sequencing_group.name,
+            bucket_name=bucket_name,
+        )
+
+        return self.make_outputs(
+            target=sequencing_group,
+            data=outputs,
+            jobs=ica_prep_job,
+        )
+
+
+@stage
+class UploadDataToIca(SequencingGroupStage):
+    def expected_outputs(self, sequencing_group: SequencingGroup) -> cpg_utils.Path:
+        sg_bucket: cpg_utils.Path = sequencing_group.dataset.prefix()
+        return sg_bucket / GCP_FOLDER_FOR_ICA_PREP / f'{sequencing_group.name}_fids.json'
+
+    def queue_jobs(self, sequencing_group: SequencingGroup, inputs: StageInput) -> StageOutput | None:  # noqa: ARG002
+        output: cpg_utils.Path = self.expected_outputs(sequencing_group=sequencing_group)
+
+        upload_job: BashJob = upload_data_to_ica.upload_data_to_ica(
+            sequencing_group=sequencing_group,
+            ica_cli_setup=ICA_CLI_SETUP,
+            output=str(output),
+        )
+
+        return self.make_outputs(
+            target=sequencing_group,
+            data=output,
+            jobs=upload_job,
+        )
+
+
+@stage(
+    required_stages=[PrepareIcaForDragenAnalysis, UploadDataToIca],  # type: ignore  # noqa: PGH003
+)
+class ManageDragenPipeline(CohortStage):
+    """
+    Due to the nature of the Dragen pipeline and stage dependencies, we need to run, monitor and cancel the pipeline in the same stage.
+
+    This stage handles the following tasks:
+    1. Cancels a previous pipeline running on ICA if requested.
+        - Set the `cancel_cohort_run` flag to `true` in the config and the stage will read the pipeline ID from the JSON file and cancel it.
+    2. Resumes monitoring a previous pipeline run if it was interrupted.
+        - Set the `monitor_previous` flag to `true` in the config. This will read the pipeline ID from the JSON file and monitor it.
+    3. Initiates a new Dragen pipeline run if no previous run is found or if resuming is not requested.
+    4. Monitors the progress of the Dragen pipeline run.
+    """  # noqa: E501
+
+    def expected_outputs(
+        self,
+        cohort: Cohort,
+    ) -> dict[str, cpg_utils.Path]:
+        sg_bucket: cpg_utils.Path = cohort.dataset.prefix()
+        return (
+            {
+                f'{sequencing_group.name}_success': sg_bucket
+                / GCP_FOLDER_FOR_RUNNING_PIPELINE
+                / f'{sequencing_group.name}_pipeline_success.json'
+                for sequencing_group in cohort.get_sequencing_groups()
+            }
+            | {
+                f'{sequencing_group.name}_pipeline_id': sg_bucket
+                / GCP_FOLDER_FOR_RUNNING_PIPELINE
+                / f'{sequencing_group.name}_pipeline_id.txt'
+                for sequencing_group in cohort.get_sequencing_groups()
+            }
+            | {f'{cohort.name}_errors': sg_bucket / GCP_FOLDER_FOR_RUNNING_PIPELINE / f'{cohort.name}_errors.log'}
+        )
```

Comment:
```suggestion
        prefix: cpg_utils.Path = sg_bucket / GCP_FOLDER_FOR_RUNNING_PIPELINE
        results: dict[str, cpg_utils.Path] = {f'{cohort.name}_errors': prefix / f'{cohort.name}_errors.log'}
        for sequencing_group in cohort.get_sequencing_groups():
            results |= {
                f'{sequencing_group.name}_success': prefix / f'{sequencing_group.name}_pipeline_success.json',
                f'{sequencing_group.name}_pipeline_id': prefix / f'{sequencing_group.name}_pipeline_id.txt',
            }
        return results
```

This syntax is an affront to god and man

Reply:
You work with what you've got I guess